### PR TITLE
Fix generation and metageneration

### PR DESCRIPTION
**Pull request by mtusnio**
I was getting this odd errors where preconditions were failing on requests sent to the [fake GCS server](https://github.com/fsouza/fake-gcs-server/):

```
time=2024-09-23T12:27:05.260Z level=INFO msg="10.244.0.8 - - [23/Sep/2024:12:27:05 +0000] \"POST /upload/storage/v1/b/tf-controller-lock/o?alt=json&ifGenerationMatch=0&name=helloworld-oci-auto-approve.lock&prettyPrint=false&projection=full&uploadType=multipart HTTP/1.1\" 412 149\n"
```

This does not seem to be an issue when talking to GCS from what I understand, considering the lock is in production, however it does not seem like the preconditions are set correctly. 

According to [docs](https://cloud.google.com/storage/docs/metadata#generation-number) the metagenerations start with 1, hence I changed that in the default new lock.

They also mention that "The metageneration property has no meaning without the generation property and should be used only in conjunction with it; it's meaningless to compare metadata generations of two object versions.", and that both can be used in preconditions, which seems to fit our use case. Added that in there to pair it up with metageneration, which seems to have solved the GCS server responding only with 412 (it still drops some 412s, but seems like the client knows how to recover now) to any refresh of locks.

Not sure if we need changes in tests for this apart from testing the field is not defaulting to zero? Seems like that comes from the the attributes, so as long as it gets changed we should be good?

### CODE REVIEW

File: leadership.go [hunk 1]
```diff
@@ -128,6 +132,7 @@ func (l *Lock) RefreshLock(ctx context.Context) error {
 
 	l.refreshFailures = 0
 	l.latestMetadataGeneration = attrs.Metageneration
+	l.latestGeneration = attrs.Generation
```
Comment (wjam) [inline leadership.go:135]: Given we don't write anything to the object - we only write to the metadata - wouldn't the generation never change?

File: leadership.go [hunk 2]
```diff
@@ -48,7 +50,7 @@ func NewLock(bucket *storage.BucketHandle, id, path string, ttl time.Duration, l
 		logger:                   logContext,
 		mutex:                    sync.Mutex{},
 		refreshMetadata:          false,
-		latestMetadataGeneration: 0,
+		latestMetadataGeneration: 1,
```
Comment (wjam) [inline leadership.go:53]: Why does this need to be changed, given that it'll get overwritten when `createLock` is called?

File: leadership.go [hunk 3]
```diff
@@ -109,7 +111,9 @@ func (l *Lock) RefreshLock(ctx context.Context) error {
 	l.logger(ctx).Info("Refreshing lock", "path", l.path)
 
 	attrs, err := l.bucket.Object(l.path).
-		If(storage.Conditions{MetagenerationMatch: l.latestMetadataGeneration}).
+		If(storage.Conditions{
+			GenerationMatch:     l.latestGeneration,
+			MetagenerationMatch: l.latestMetadataGeneration}).
```
Comment (matpimenta) [inline leadership.go:116]: Do we need to have any conditions on metadata? 
I understand that we only care about the GenerationMatch, which should always match `0`.

File: leadership.go (hunk 1 above)
Comment (matpimenta) [inline leadership.go:135]: I think the behaviour here is that `l.latestGeneration` should always be `0`. If it's something else, it means that someone updated the lock, which should happen.

**Comment by matpimenta** [comment]
Reference blog post https://www.joyfulbikeshedding.com/blog/2021-05-19-robust-distributed-locking-algorithm-based-on-google-cloud-storage.html

File: leadership.go (hunk 1 above)
Comment (matpimenta) [reply leadership.go]: My suggestion was incorrect, so resolving the comment

**Review by wjam — APPROVED** [review]
